GLES2 Batching - Fix custom shaders reading VERTEX

lawnjelly opened the pull request:
In situations where custom canvas shaders read VERTEX values, they could read incorrect global positions rather than local positions where batching had baked item transforms. This PR prevents joining items that read VERTEX built in in shaders, and thus prevents this situation arising (as unjoined items will not bake transforms).

Fixes #38664.

## Notes
* I've changed the 'prevent' bools to bitflags as it probably makes more sense if we are using more than one.
* As such any reviewer should double check I haven't mucked up the bitwise logic(!!) :grinning:

lawnjelly committed:
GLES2 Batching - Fix custom shaders reading VERTEX

In situations where custom canvas shaders read VERTEX values, they could read incorrect global positions rather than local positions where batching had baked item transforms. This PR prevents joining items that read VERTEX built in in shaders, and thus prevents this situation arising (as unjoined items will not bake transforms).

clayjohn reviewed (approved):
Changes look good! Too bad it creates one more situation where batching isn't possible.

lawnjelly commented:
> Changes look good! Too bad it creates one more situation where batching isn't possible.

Well, this may be only temporary, just to get it working. In the long run, we can consider adding a larger vertex format for such cases, which contain 2 sets of color, and 2 positions. These should batch successfully and not break these custom shaders.

In fact I doubt they will be _that_ much more expensive than the smaller vertex format. If I had to guess I'd say they might run at 90% speed or so.

lawnjelly commented:
Actually I've found another case which also needs to be fixed (isn't seen in the min project):

It will also batch within the item, with software transform of the extra matrix, which can also cause a similar problem with VERTEX reading shaders. I'm just looking at fixing it now. I can either fix it in this PR or do a separate one, as the fixes are independent (although related).

clayjohn commented:
You can just add it in here.

lawnjelly committed:
GLES2 Batching - Prevent VERTEX baking within items in custom shaders

In addition to prevent item joins when VERTEX reads are present in a custom shader, it is also necessary to prevent baking extra matrices (extra transforms) WITHIN items, because these can also report incorrect results.

lawnjelly commented:
Ok that should do the trick. I've added as a separate commit to make it more obvious for reviewing, I can squash these if they are okay.

I've changed to just passing the bitflags (prevent color baking, prevent vertex baking) directly to bdata, as it now no longer makes a lot of sense to pass around several bools.

akien-mga commented:
Thanks!
